kernel/gic/src/gic/mod.rs

```diff
@@ -349,72 +332,88 @@ impl ArmGic {
     pub fn send_ipi(&mut self, int_num: InterruptNumber, target: IpiTargetCpu) {
         assert!(int_num < 16, "IPIs must have a number below 16 on ARMv8");
 
-        match self {
-            Self::V2(v2) => dist_interface::send_ipi_gicv2(&mut v2.distributor, int_num, target),
-            Self::V3( _) => cpu_interface_gicv3::send_ipi(int_num, target),
+        if let Self::V3 { .. } = self {
+            cpu_interface_gicv3::send_ipi(int_num, target)
+        } else {
+            // we don't have access to the distributor... code would be:
+            // dist_interface::send_ipi_gicv2(&mut dist_regs, int_num, target)
+            // workaround: caller could check is this must be done in the dist
+            // and then get the SystemInterruptController and call a dedicated
+            // method on it, like `sys_ctlr.send_ipi_gicv2()`
+
+            panic!("GICv2 doesn't support sending IPIs (need distributor)");
         }
     }
 
     /// Acknowledge the currently serviced interrupt
     /// and fetches its number
     pub fn acknowledge_interrupt(&mut self) -> (InterruptNumber, Priority) {
         match self {
-            Self::V2(v2) => cpu_interface_gicv2::acknowledge_interrupt(&mut v2.processor),
-            Self::V3( _) => cpu_interface_gicv3::acknowledge_interrupt(),
+            Self::V2 { registers, .. } => registers.acknowledge_interrupt(),
+            Self::V3 { .. } => cpu_interface_gicv3::acknowledge_interrupt(),
         }
     }
 
     /// Performs priority drop for the specified interrupt
     pub fn end_of_interrupt(&mut self, int: InterruptNumber) {
         match self {
-            Self::V2(v2) => cpu_interface_gicv2::end_of_interrupt(&mut v2.processor, int),
-            Self::V3( _) => cpu_interface_gicv3::end_of_interrupt(int),
+            Self::V2 { registers, .. } => registers.end_of_interrupt(int),
+            Self::V3 { .. } => cpu_interface_gicv3::end_of_interrupt(int),
         }
     }
 
-    /// Will that interrupt be forwarded by the distributor?
+    /// Will that interrupt be received by this CPU?
     pub fn get_interrupt_state(&self, int: InterruptNumber) -> Enabled {
-        match (int, self) {
-            (0..=31, Self::V3(v3)) => {
-                let i = get_current_cpu_redist_index();
-                redist_interface::is_sgippi_enabled(&v3.redistributors[i].redist_sgippi, int)
-            },
-            (_, this) => dist_interface::is_spi_enabled(this.distributor(), int),
+        assert!(int < 32);
```

Comment:
all of these assertions are good, but we should make two improvements to each of them:
1. add a string message that is specific to each function, explaining why it failed.
2. Add docs to the function describing the conditions under which it will panic. You can do that either using a section header `/// # Panic` or as a description of the `int` argument, i.e., that is must be under 32.